src/openrct2/actions/LandSetRightsAction.hpp

```diff
@@ -0,0 +1,231 @@
+                    map_invalidate_tile(loc.x, loc.y, baseHeight, baseHeight + 16);
+                }
+                return res;
+            case LandSetRightSetting::SetConstructionRightsForSale:
+                if (isExecuting)
+                {
+                    surfaceElement->SetOwnership(surfaceElement->GetOwnership() | OWNERSHIP_CONSTRUCTION_RIGHTS_AVAILABLE);
+                    uint16_t baseHeight = surfaceElement->base_height * 8;
+                    map_invalidate_tile(loc.x, loc.y, baseHeight, baseHeight + 16);
+                }
+                return res;
+            case LandSetRightSetting::SetOwnershipWithChecks:
+            {
+                if (_ownership == surfaceElement->GetOwnership())
+                {
+                    return res;
+                }
+
+                TileElement* tileElement = map_get_first_element_at(loc.x / 32, loc.y / 32);
+                do
+                {
+                    if (tileElement->GetType() != TILE_ELEMENT_TYPE_ENTRANCE)
+                        continue;
+
+                    if (tileElement->AsEntrance()->GetEntranceType() != ENTRANCE_TYPE_PARK_ENTRANCE)
+                        continue;
+
+                    // Do not allow ownership of park entrance.
+                    if (_ownership == OWNERSHIP_OWNED || _ownership == OWNERSHIP_AVAILABLE)
+                        return res;
+
+                    // Allow construction rights available / for sale on park entrances on surface.
+                    // There is no need to check the height if _ownership is 0 (unowned and no rights available).
+                    if (_ownership == OWNERSHIP_CONSTRUCTION_RIGHTS_OWNED
+                        || _ownership == OWNERSHIP_CONSTRUCTION_RIGHTS_AVAILABLE)
+                    {
+                        if (tileElement->base_height - 3 > surfaceElement->base_height
+                            || tileElement->base_height < surfaceElement->base_height)
+                            return res;
+                    }
+                } while (!(tileElement++)->IsLastForTile());
+
+                res->Cost = gLandPrice;
+                if (isExecuting)
+                {
+                    if (_ownership != OWNERSHIP_UNOWNED)
+                    {
+                        gPeepSpawns.erase(
+                            std::remove_if(
+                                gPeepSpawns.begin(), gPeepSpawns.end(),
+                                [x = loc.x, y = loc.y](const auto& spawn) {
+                                    return floor2(spawn.x, 32) == x && floor2(spawn.y, 32) == y;
+                                }),
+                            gPeepSpawns.end());
+                    }
+                    surfaceElement->SetOwnership(_ownership);
+                    update_park_fences_around_tile(loc);
+                    gMapLandRightsUpdateSuccess = true;
+                }
+                return res;
+            }
+            default:
```

Comment:
Should be done in Query

Reply:
This is the Query and Execute. As above.

Reply:
Execute is never called without Query first unless one would specifically invoke Execute without the game action framework so its just doing it twice at the moment.

Reply:
I don't really see the point in adding extra logic so that it can only execute during Query.

Reply:
What I meant is it has to be done only in Query as long no one directly uses Execute, but its fine this way as well.